docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,389 @@
+  
+  Spec   IPAddressClaimSpec   `json:"spec,omitempty"`
+  Status IPAddressClaimStatus `json:"status,omitempty"`
+}
+```
+
+##### IPAddress
+
+IPAddress resources are created to fulfil an IPAddressClaim. They are created by the IPAM provider that reconciles the claim.
+
+IPAddresses are immutable.
+
+```go
+// IPAddressSpec describes an IPAddress
+type IPAddressSpec struct {
+  // Claim is a reference to the claim this IPAddress was created for.
+  Claim LocalObjectReference `json:"claim,omitempty"`
+ 
+  // Pool is a reference to the pool that this IPAddress was created from.
+  Pool LocalObjectReference `json:"pool,omitempty"`
+ 
+  // Address is the IP address.
+  Address string `json:"address"`
+ 
+  // Prefix is the prefix of the address.
+  Prefix int `json:"prefix,omitempty"`
+ 
+  // Gateway is the network gateway of network the address is from.
+  Gateway string `json:"gateway,omitempty"`
+}
+ 
+// IPAddress is a representation of an IP Address that was allocated from an IP Pool.
+type IPAddress struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec IPAddressSpec `json:"spec,omitempty"`
+}
+```
+
+##### New Reference Type
+
+```go
+type LocalObjectReference struct {
+  Group string
+  Kind string
+  Name string
+}
+```
+
+#### Implementing an IPAM Provider
+
+IPAM providers have to provide an IP Pool resource. It serves two purposes: Selecting which provider to use, and configuring that provider. How configuration works exactly is up to the provider. The resource can allow to select a specific network to allocate addresses from, where different clusters might use different pools. Or there could only be a single object shared with all clusters, with the provider deciding what network to use on its own.
+
+The primary task of the IPAM provider is fulfilling IPAddressClaims, usally issued by infrastructure providers. The provider therefore has to reconcile IPAddressClaims that are referencing one of its pools. It has to allocate IPs and create IPAddress objects for each claim.
+
+A claim is fulfilled once an IPAddress object is created, and the status of the claim is updated with a reference to that IPAddress object. During the allocation process, the `.Status.Conditions` array of the IPAddressClaim should be used to reflect the status of the allocation process. Especially errors have to be visible.
+
+When an IPAddressClaim that is reconciled by a provider is deleted, the provider needs to release the IP address belonging to the claim and delete the related IPAddress object. To do so a Finalzier should be added to the claim to prevent its deletion until the address is released.
+
+IPAddress objects (as well as claims) are immutable. For each IPAddressClaim, one corresponding IPAddress object gets created. As long as the claim exists that IPAddress object must not change. When an IPAddress object is deleted while its claim still exists there are two options:
```

Comment:
Probably already clear but I assume only the spec's are immutable.

Reply:
Yes.